packages/bug-detectors/internal/path-traversal.ts

```diff
@@ -0,0 +1,213 @@
+			"rmdirSync",
+			"stat",
+			"statfs",
+			"statfsSync",
+			"statSync",
+			"symlink",
+			"symlinkSync",
+			"truncate",
+			"truncateSync",
+			"unlink",
+			"unlinkSync",
+			"unwatchFile",
+			"utimes",
+			"utimesSync",
+			"watch",
+			"watchFile",
+			"writeFile",
+			"writeFileSync",
+		],
+	},
+	{
+		moduleName: "fs/promises",
+		functionNames: [
+			"access",
+			"appendFile",
+			"chmod",
+			"chown",
+			"copyFile",
+			"cp",
+			"lchmod",
+			"lchown",
+			"link",
+			"lstat",
+			"lutimes",
+			"mkdir",
+			"open",
+			"opendir",
+			"readFile",
+			"readlink",
+			"readdir",
+			"realpath",
+			"rename",
+			"rm",
+			"rmdir",
+			"stat",
+			"statfs",
+			"symlink",
+			"truncate",
+			"unlink",
+			"utimes",
+			"watch",
+			"writeFile",
+		],
+	},
+	// path.join() can have any number of strings as inputs. Internally, it uses path.normalize(), which we hook here.
+	{
+		moduleName: "path",
+		functionNames: ["normalize", "resolve"],
+	},
+];
+
+for (const module of modulesToHook) {
+	for (const functionName of module.functionNames) {
+		const beforeHook = (
+			thisPtr: unknown,
+			params: unknown[],
+			hookId: number
+		) => {
+			if (params === undefined || params.length === 0) {
+				return;
+			}
+			// The first argument of the original function is typically
+			// a path or a file name.
+			const firstArgument = params[0] as string;
+			if (firstArgument.includes(goal)) {
+				reportFinding(
+					`Path Traversal in ${functionName}(): called with '${firstArgument}'`
+				);
+			}
+			guideTowardsContainment(firstArgument, goal, hookId);
+		};
+
+		registerBeforeHook(functionName, module.moduleName, false, beforeHook);
+	}
+}
+
+// Some functions have two arguments that can be used for path traversal.
+const functionsWithTwoTargets = [
+	{
+		moduleName: "fs/promises",
+		functionNames: ["copyFile", "cp", "link", "rename", "symlink"],
+	},
+	{
+		moduleName: "fs",
+		functionNames: [
+			"copyFile",
+			"copyFileSync",
+			"cp",
+			"cpSync",
+			"link",
+			"linkSync",
+			"rename",
+			"renameSync",
+			"symlink",
+			"symlinkSync",
+		],
+	},
+];
+
+for (const module of functionsWithTwoTargets) {
+	for (const functionName of module.functionNames) {
+		const beforeHook = (
+			thisPtr: unknown,
+			params: unknown[],
+			hookId: number
+		) => {
+			if (params === undefined || params.length < 2) {
+				return;
+			}
+			// The first two arguments are paths.
+			const firstArgument = params[0] as string;
+			const secondArgument = params[1] as string;
+			if (firstArgument.includes(goal) || secondArgument.includes(goal)) {
+				reportFinding(
+					`Path Traversal in ${functionName}(): called with '${firstArgument}'` +
+						` and '${secondArgument}'`
+				);
+			}
+			guideTowardsContainment(firstArgument, goal, hookId);
+			// We don't want to confuse the fuzzer with the same hookId (used as a program counter (PC)),
+			// so we increment it.
+			guideTowardsContainment(secondArgument, goal, hookId + 1);
```

Comment:
The alternative is to use a random id, in which case next id is as good as any other. But maybe I can add a function like e.g. `nextFreeId` that gives me a free id if there is one available, or a random one if there is not.

Reply:
Thinking a bit about this, implementing this the way I see it requires changes in the HookingManager, so that each hook tracks its own id. Then we could just ask the manager to give us an unused id, and blocking it for the future hooks. Since there are only 512 fake PCs available, the manager should reset the block and start counting from 0 again, once all 512 PCs were taken. I imagine that more than 512 functions might need to be hooked.
It's really too bad that there are only 512 slots, because they are also shared with comparison instrumentation in ´compareHooks.ts`. We might as well just take a random fake PC for each new hook function as well!

Reply:
Looking at it again, the hooks of the instrumented functions actually use callSiteId() to generate their IDs.
I exported this from the hook manager and added to the path traversal.

Reply:
Couldn't this interfere with reproducibility, though? Now hook ids could randomly clash.